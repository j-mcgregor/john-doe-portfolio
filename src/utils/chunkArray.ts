export function chunkArray<T>(array: T[], n: number): Array<T[]> {
    const copy: T[] = [...array]
    const result: Array<T[]> = []

    while (copy.length) {
        const num: number = Math.ceil(copy.length / n--)
        const item: T[] = copy.splice(0, num)
        result.push(item)
    }
    return result
}
