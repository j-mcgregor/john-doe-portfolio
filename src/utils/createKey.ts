// RULES
// 1. Each string should have '-' not ' '
// 2. Should be lower case
// 3. Should join to the index

const formatString = (str: string) =>
    str.toLowerCase().replace(/[^0-9a-zA-Z]+/g, '-')

export default (value: string | string[], index = 0) => {
    if (Array.isArray(value)) {
        return `${value.map(formatString).join('-')}-${index}`
    } else {
        return `${formatString(value)}-${index}`
    }
}
