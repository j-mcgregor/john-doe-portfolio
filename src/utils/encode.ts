import { stringify } from 'query-string'
export default function <T>(data: T) {
    const formData = new FormData()
    Object.keys(data).forEach(k => {
        formData.append(k, data[k])
    })
    return stringify(formData)
}
