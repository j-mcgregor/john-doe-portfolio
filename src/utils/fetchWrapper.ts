/* eslint-disable @typescript-eslint/no-explicit-any */
import axios from 'axios'

export const fetchWrapper = async <T, P>(
    url: string,
    data: string,
    headers: any
): Promise<T> => {
    try {
        const res = await axios({ method: 'POST', url, data, headers })
        return res.data
    } catch (error) {
        return error
    }
}
