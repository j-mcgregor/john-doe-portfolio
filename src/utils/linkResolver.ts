interface LinkResolverProp {
    type: string
    uid: string
}

export const linkResolver = (doc: LinkResolverProp) => {
    if (doc.type === 'page') {
        return `/${doc.uid}`
    }

    return '/'
}
