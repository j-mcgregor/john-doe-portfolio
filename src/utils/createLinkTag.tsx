import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faFacebook, faLinkedinIn } from '@fortawesome/free-brands-svg-icons'
import { faEnvelope } from '@fortawesome/free-regular-svg-icons'
import { faPhone } from '@fortawesome/free-solid-svg-icons'
import createKey from './createKey'

export const createLink = (type: string, url: string, target = false) => {
    const targetProps = target
        ? { target: '_blank', rel: 'noopener noreferrer' }
        : {}

    switch (type) {
        case 'Facebook':
            return (
                <a href={url} {...targetProps} key={createKey(type)}>
                    <FontAwesomeIcon icon={faFacebook} />
                </a>
            )
        case 'LinkedIn':
            return (
                <a href={url} {...targetProps} key={createKey(type)}>
                    <FontAwesomeIcon icon={faLinkedinIn} />
                </a>
            )
        case 'Mail':
            return (
                <a href={url} {...targetProps} key={createKey(type)}>
                    <FontAwesomeIcon icon={faEnvelope} />
                </a>
            )
        case 'Phone':
            return (
                <a href={url} {...targetProps} key={createKey(type)}>
                    <FontAwesomeIcon icon={faPhone} />
                </a>
            )
        default:
            return null
    }
}
