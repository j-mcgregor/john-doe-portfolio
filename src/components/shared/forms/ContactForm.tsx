import React, { useState } from 'react'
import axios from 'axios'
import Notification from '../Notification'
import TextInput from './inputs/TextInput'
import TextAreaInput from './inputs/TextAreaInput'
import Button from '../buttons/Button'
import CheckboxInput from './inputs/CheckboxInput'
import encode from '../../../utils/encode'

export interface SubmitProps {
    firstName: string
    lastName: string
    email: string
    message: string
    subscribe: boolean
}
export interface ContactFormProps {
    onFirstNameChange?: (firstName: string) => void
    onLastNameChange?: (lastName: string) => void
    onEmailChange?: (email: string) => void
    onMessageChange?: (message: string) => void
    onSubscribe?: (subscribe: boolean) => void
    onSubmit?: (props: SubmitProps) => void
    handleCallback?: (callback: Function, arg: boolean | null) => void
    subscribe?: boolean
}

const ContactForm = (props: ContactFormProps) => {
    const [firstName, setFirstName] = useState<string>('')
    const [lastName, setLastName] = useState<string>('')
    const [email, setEmail] = useState<string>('')
    const [message, setMessage] = useState<string>('')
    const [subscribe, setSubscribe] = useState<boolean>(false)
    const [status, setStatus] = useState<true | false | null>(null)
    const [, setLoading] = useState<boolean>(false)

    const handleChange = (
        e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>
    ) => {
        const { name, value } = e.currentTarget
        if (name === 'firstName') {
            setFirstName(value)
            props.onFirstNameChange && props.onFirstNameChange(value)
        } else if (name === 'lastName') {
            setLastName(value)
            props.onLastNameChange && props.onLastNameChange(value)
        } else if (name === 'email') {
            setEmail(value)
            props.onEmailChange && props.onEmailChange(value)
        } else if (name === 'message') {
            setMessage(value)
            props.onMessageChange && props.onMessageChange(value)
        }
    }

    const handleCheckbox = (e: React.ChangeEvent<HTMLInputElement>) => {
        const { checked } = e.target
        setSubscribe(checked)
        props.onSubscribe && props.onSubscribe(checked)
    }

    const handleSubmit = async (e: React.FormEvent) => {
        e.preventDefault()
        const data = { firstName, lastName, email, message, subscribe }
        const headers = { 'Content-Type': 'application/x-www-form-urlencoded' }

        setLoading(true)
        try {
            const res = await axios({
                method: 'POST',
                url: 'https://john-doe-portfolio.netlify.app/',
                data: encode({ 'form-name': 'contact-form', ...data }),
                headers,
            })
            console.log(res)
            setStatus(true)
            props.onSubmit && props.onSubmit(data)
        } catch (e) {
            setStatus(false)
        } finally {
            setLoading(false)
        }
    }

    return (
        <form
            name="contact-form"
            onSubmit={handleSubmit}
            data-testid="contact-form"
            data-netlify="true"
            data-netlify-honeypot="bot-field"
        >
            <input type="hidden" name="form-name" value="contact-form" />
            <div className="row">
                <TextInput
                    type="text"
                    placeholder="First Name"
                    name="firstName"
                    classNames={['col-md-6']}
                    value={firstName}
                    onChange={handleChange}
                    required
                />
                <TextInput
                    type="text"
                    placeholder="Last Name"
                    name="lastName"
                    classNames={['col-md-6']}
                    value={lastName}
                    onChange={handleChange}
                    required
                />
            </div>
            <TextInput
                type="email"
                placeholder="Email"
                name="email"
                value={email}
                onChange={handleChange}
                required
            />
            <TextAreaInput
                placeholder="Message"
                name="message"
                value={message}
                onChange={handleChange}
                required
            />
            <CheckboxInput
                isChecked={subscribe}
                onChange={handleCheckbox}
                name="subscribe"
                label="Subscribe to newsletter?"
            />
            <Button type="submit" value="SEND MESSAGE" className="mt1" />
            {status === true ? (
                <Notification
                    type="success"
                    message="Success!"
                    onClick={() => setStatus(null)}
                />
            ) : null}
            {status === false ? (
                <Notification
                    type="failure"
                    message="Oops! Something went wrong"
                    onClick={() => setStatus(null)}
                />
            ) : null}
        </form>
    )
}

export default ContactForm
