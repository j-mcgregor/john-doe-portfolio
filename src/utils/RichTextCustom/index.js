import React from 'react'
import { RichText } from 'prismic-reactjs'
import { linkResolver } from './linkResolver'
import { htmlSerializer } from './htmlSerializer'

const RichTextCustom = ({ render }) => (
    <RichText
        render={render}
        linkResolver={linkResolver}
        htmlSerializer={htmlSerializer}
    />
)

export default RichTextCustom
