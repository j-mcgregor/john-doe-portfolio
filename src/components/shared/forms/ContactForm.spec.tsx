import React from 'react'
import { act } from 'react-dom/test-utils'
import { mocked } from 'ts-jest/utils'
import { render, fireEvent } from '@testing-library/react'
import ContactForm, { ContactFormProps } from './ContactForm'
import { fetchWrapper } from '../../../utils/fetchWrapper'

jest.mock('../../../utils/fetchWrapper')

const renderContactForm = (props: Partial<ContactFormProps> = {}) => {
    const defaultProps: ContactFormProps = {
        onFirstNameChange: () => {},
        onLastNameChange: () => {},
        onEmailChange: () => {},
        onMessageChange: () => {},
        onSubscribe: () => {},
        onSubmit: () => {},
        handleCallback: (cb: Function, arg: boolean | null) => cb(arg),
        subscribe: false,
    }
    return render(<ContactForm {...defaultProps} {...props} />)
}

describe('ContactForm', () => {
    let mockedFetch

    beforeEach(() => {
        jest.clearAllMocks()
        mockedFetch = mocked(fetchWrapper)
    })
    it('should render the contact form', () => {
        const utils = renderContactForm()
        const form = utils.getByTestId('contact-form')

        expect(form).toHaveFormValues({
            firstName: '',
            lastName: '',
            email: '',
            message: '',
            subscribe: false,
        })
        expect(utils.baseElement).toMatchSnapshot()
    })

    it('should fill in the firstName field', async () => {
        // ASSEMBLE
        const onFirstNameChange = jest.fn()
        const utils = renderContactForm({ onFirstNameChange })
        const firstName = utils.getByTestId('firstName')

        // ACT
        fireEvent.change(firstName, { target: { value: 'Bruce' } })

        // ASSERT
        expect(onFirstNameChange).toHaveBeenCalledWith('Bruce')
    })

    it('should fill in the lastName field', async () => {
        // ASSEMBLE
        const onLastNameChange = jest.fn()
        const utils = renderContactForm({ onLastNameChange })
        const lastName = utils.getByTestId('lastName')

        // ACT
        fireEvent.change(lastName, { target: { value: 'Wayne' } })

        // ASSERT
        expect(onLastNameChange).toHaveBeenCalledWith('Wayne')
    })

    it('should fill in the email field', async () => {
        // ASSEMBLE
        const onEmailChange = jest.fn()
        const utils = renderContactForm({ onEmailChange })
        const email = utils.getByTestId('email')

        // ACT
        fireEvent.change(email, { target: { value: 'not@batman.com' } })

        // ASSERT
        expect(onEmailChange).toHaveBeenCalledWith('not@batman.com')
    })

    it('should fill in the message field', async () => {
        // ASSEMBLE
        const onMessageChange = jest.fn()
        const utils = renderContactForm({ onMessageChange })
        const message = utils.getByTestId('message')

        // ACT
        fireEvent.change(message, { target: { value: 'Im Batman!!' } })

        // ASSERT
        expect(onMessageChange).toHaveBeenCalledWith('Im Batman!!')
    })

    it('should toggle the subscribe field', async () => {
        const onSubscribe = jest.fn()
        const utils = renderContactForm({ onSubscribe, subscribe: false })
        const subscribe = utils.getByTestId('subscribe')

        fireEvent.click(subscribe)
        expect(onSubscribe).toHaveBeenCalledWith(true)
        fireEvent.click(subscribe)
        expect(onSubscribe).toHaveBeenCalledWith(false)
    })

    it('should submit with firstName, lastName, email, message and subscribe', async () => {
        // ASSEMBLE
        const onSubmit = jest.fn()
        const handleCallback = jest.fn()

        const utils = renderContactForm({
            subscribe: false,
            onSubmit,
            handleCallback,
        })

        const firstName = utils.getByTestId('firstName')
        const lastName = utils.getByTestId('lastName')
        const email = utils.getByTestId('email')
        const message = utils.getByTestId('message')
        const subscribe = utils.getByTestId('subscribe')
        const submit = utils.getByTestId('SEND MESSAGE')

        mockedFetch.mockImplementationOnce(() => Promise.resolve())

        // ACT
        fireEvent.change(firstName, { target: { value: 'Clark' } })
        fireEvent.change(lastName, { target: { value: 'Kent' } })
        fireEvent.change(email, {
            target: { value: 'lastson@krypton.com' },
        })
        fireEvent.change(message, { target: { value: 'I like the sun' } })
        fireEvent.click(subscribe)

        await act(async () => {
            fireEvent.click(submit)
        })

        // ASSERT
        expect(onSubmit).toHaveBeenCalledWith({
            firstName: 'Clark',
            lastName: 'Kent',
            email: 'lastson@krypton.com',
            message: 'I like the sun',
            subscribe: true,
        })
    })

    it('should show <Notification type="success" /> if status is true', async () => {
        // ASSEMBLE
        const utils = renderContactForm()

        const submit = utils.getByTestId('SEND MESSAGE')

        mockedFetch.mockImplementationOnce(() => Promise.resolve())

        // ACT
        await act(async () => {
            fireEvent.click(submit)
        })

        // ASSERT
        expect(utils.getByTestId('notify-success')).toBeDefined()
    })

    it('should show <Notification type="failure" /> if status is false', async () => {
        // ASSEMBLE
        const utils = renderContactForm()

        const submit = utils.getByTestId('SEND MESSAGE')

        mockedFetch.mockImplementationOnce(() => Promise.reject())

        // ACT
        await act(async () => {
            fireEvent.click(submit)
        })

        // ASSERT
        expect(utils.getByTestId('notify-failure')).toBeDefined()
    })

    it('should remove the success notification when clicked', async () => {
        // ASSEMBLE
        const utils = renderContactForm()
        const submit = utils.getByTestId('SEND MESSAGE')
        mockedFetch.mockImplementationOnce(() => Promise.resolve())

        // ACT
        await act(async () => {
            fireEvent.click(submit)
        })

        // ASSERT
        expect(utils.getByTestId('notify-success')).toBeDefined()
        fireEvent.click(utils.getByTestId('notify-close'))
        // .query returns null if not found, instead of throwing error
        expect(utils.queryByTestId('notify-success')).toBeNull()
    })

    it('should remove the failure notification when clicked', async () => {
        // ASSEMBLE
        const utils = renderContactForm()
        const submit = utils.getByTestId('SEND MESSAGE')
        mockedFetch.mockImplementationOnce(() => Promise.reject())

        // ACT
        await act(async () => {
            fireEvent.click(submit)
        })

        // ASSERT
        expect(utils.getByTestId('notify-failure')).toBeDefined()
        fireEvent.click(utils.getByTestId('notify-close'))
        // .query returns null if not found, instead of throwing error
        expect(utils.queryByTestId('notify-failure')).toBeNull()
    })
})
