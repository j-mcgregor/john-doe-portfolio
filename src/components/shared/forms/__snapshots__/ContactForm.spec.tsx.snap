// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm should render the contact form 1`] = `
<body>
  <div>
    <form
      data-testid="contact-form"
    >
      <input
        name="form-name"
        type="hidden"
        value="contact-form"
      />
      <div
        class="row"
      >
        <div
          class="form-group col-md-6"
        >
          <input
            aria-label="firstName"
            class="form-control"
            data-testid="firstName"
            name="firstName"
            placeholder="First Name"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group col-md-6"
        >
          <input
            aria-label="lastName"
            class="form-control"
            data-testid="lastName"
            name="lastName"
            placeholder="Last Name"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="form-group "
      >
        <input
          aria-label="email"
          class="form-control"
          data-testid="email"
          name="email"
          placeholder="Email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <textarea
          aria-label="message"
          class="form-control"
          data-testid="message"
          name="message"
          placeholder="Message"
          required=""
        />
      </div>
      <div
        class="form-group flex flex-row"
      >
        <input
          aria-label="subscribe"
          data-testid="subscribe"
          name="subscribe"
          type="checkbox"
        />
        <label
          for="subscribe"
        >
          Subscribe to newsletter?
        </label>
      </div>
      <button
        class="btn mt1"
        data-testid="SEND MESSAGE"
        type="submit"
      >
        SEND MESSAGE
      </button>
    </form>
  </div>
</body>
`;
